stl/inc/ranges

```diff
@@ -22,15 +25,461 @@ _STL_DISABLE_CLANG_WARNINGS
 
 _STD_BEGIN
 namespace ranges {
-    // Much machinery defined in <xutility>
+    // MUCH machinery defined in <xutility>
 
     // clang-format off
     // CONCEPT ranges::viewable_range
     template <class _Rng>
     concept viewable_range = range<_Rng>
         && (borrowed_range<_Rng> || view<remove_cvref_t<_Rng>>);
+
+    // CLASS TEMPLATE ranges::_Semiregular_box
+    // clang-format off
+    template <class _Ty>
+    concept _Copy_constructible_object = copy_constructible<_Ty> && is_object_v<_Ty>;
+    // clang-format on
+
+#if 0 // TRANSITION, VSO-1174090
+    template <_Copy_constructible_object _Ty>
+    class _Semiregular_box {
+    public:
+        constexpr _Semiregular_box() noexcept : _Dummy{}, _Engaged{false} {}
+        constexpr _Semiregular_box() noexcept(is_nothrow_default_constructible_v<_Ty>) requires constructible_from<_Ty>
+            : _Val(), _Engaged{true} {}
+
+        template <class... _Types>
+        constexpr _Semiregular_box(in_place_t, _Types&&... _Args) noexcept(
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            : _Val(_STD forward<_Types>(_Args)...), _Engaged{true} {}
+
+        ~_Semiregular_box() requires is_trivially_destructible_v<_Ty> = default;
+
+        ~_Semiregular_box() {
+            if (_Engaged) {
+                _Val.~_Ty();
+            }
+        }
+
+        _Semiregular_box(const _Semiregular_box&) requires is_trivially_copy_constructible_v<_Ty> = default;
+        _Semiregular_box(const _Semiregular_box& _That) : _Engaged{_That._Engaged} {
+            if (_Engaged) {
+                _Construct_in_place(_Val, _That._Val);
+            }
+        }
+
+        _Semiregular_box(_Semiregular_box&&) requires is_trivially_move_constructible_v<_Ty> = default;
+        _Semiregular_box(_Semiregular_box&& _That) : _Engaged{_That._Engaged} {
+            if (_Engaged) {
+                _Construct_in_place(_Val, _STD move(_That._Val));

```

Comment:
This should be `_That._Engaged`

Reply:
They're equal, but I suppose the optimizer still knows they're equal here so why not.